engine/runtime-integration-tests/src/test/java/org/enso/interpreter/runtime/progress/ProgressTest.java

```diff
@@ -0,0 +1,188 @@
+    assertEquals("geometric sequence+1", oneTimeLog.get(3).msg());
+
+    var r2 = geom.execute(2, 2.0, 0.5);
+    assertEquals("Three", 3.0, r2.asDouble(), 0.001);
+    var r3 = geom.execute(3, 2.0, 0.5);
+    assertEquals("Three and half", 3.5, r3.asDouble(), 0.001);
+
+    var fiftyTimes =
+        TestLogger.apply(
+            log,
+            () -> {
+              var r4 = geom.execute(50, 2.0, 0.5);
+              assertEquals("Got almost four", 4.0, r4.asDouble(), 0.001);
+            });
+    assertEquals("50*2 + 2: " + fiftyTimes, 102, fiftyTimes.size());
+  }
+
+  @Test
+  public void useExistingProgressFromJava() throws Exception {
+    performExistingProgressFromJavaWith(new Accumulator(1));
+  }
+
+  @Test
+  public void useExistingProgressFromJavaViaProgressInterface() throws Exception {
+    performExistingProgressFromJavaWith(new AccumulatorWithProgress(1));
+  }
+
+  /**
+   * Expecting {@code acc} to have methods {@code accumulate} (two arguments) and {@code result} (no
+   * argument).
+   */
+  private void performExistingProgressFromJavaWith(Object acc) {
+    var code =
+        """
+    from Standard.Base import Integer, Float
+    from Standard.Base.Logging import Progress
+
+    up_to n combine =
+        Progress.run "from 0 to "+n.to_text n progress->
+            loop count_down =
+                if count_down <= 0 then combine.result else
+                    combine.accumulate count_down progress
+                    @Tail_Call loop count_down-1
+
+            loop n
+    """;
+    var upTo = ctx.eval("enso", code).invokeMember(MethodNames.Module.EVAL_EXPRESSION, "up_to");
+
+    var log = LoggerFactory.getLogger("Standard.Base.Logging.Progress");
+
+    var msgs =
+        TestLogger.apply(
+            log,
+            () -> {
+              var fac5 = upTo.execute(5, acc);
+              assertEquals(120, fac5.asInt());
+            });
+
+    assertEquals("Seven messsages " + msgs, 7, msgs.size());
+    var txt = msgs.stream().map(TestLogMessage::msg).collect(Collectors.joining("\n"));
+    assertEquals(
+        "Initialize five steps. Then five `advance` calls and finally advance to finish.",
+        """
+        from 0 to 5@5
+        from 0 to 5+1
+        from 0 to 5+1
+        from 0 to 5+1
+        from 0 to 5+1
+        from 0 to 5+1
+        from 0 to 5+5""",
+        txt);
+  }
+
+  public static interface Progress {
+    public void advance(int steps);
+
+    public void log(String detail);
+  }
+
+  public static final class Accumulator {
+    private long mul;
+
+    private Accumulator(long mul) {
+      this.mul = mul;
+    }
+
+    public void accumulate(Long t, Value progress) {
+      mul *= t;
+      progress.invokeMember("advance", 1);
+    }
+
+    public Long result() {
+      return mul;
+    }
+  }
+
+  public static final class AccumulatorWithProgress {
+    // implements BiConsumer<Long, Progress>, Supplier<Long> {
```

Comment:
?

Reply:
I tried to be _type safe_ in https://github.com/enso-org/enso/pull/12163/commits/08497fb57de54070c7524f6e525d32177dfadc04#diff-6bf51b15bc0a5c0977d3598560cda370dd47d1a02454122b2489016f56513036R139, but then I found such an approach cannot be expanded to 2818c0ecb711a737aa7c82274bd16d3ed27d13db because of [generic type erasure](http://wiki.apidesign.org/wiki/Erasure) and had to comment the `implements` out.